### PR TITLE
Deprecate listDirFactory in HealObjects, rely on ListObjectsHeal

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Deprecate listDirFactory in HealObjects, rely on ListObjectsHeal

## Motivation and Context
Deprecate legacy low-performance code

## How to test this PR?
Just test background Heal() code by starting a server instance with missing either `config` or `config/iam` data. 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#fbe03e75-78c0-4df0-a342-f046141b1089) |
| mint-xl.sh | :x: [more...](#f370a751-0004-4c86-8d5d-26597800b5c4) |
| mint-large-bucket.sh | :x: [more...](#ddd6ad3c-e133-4435-8206-e64f6ee41bb7) |
| mint-compression-dist-xl.sh | :x: [more...](#6d0f7539-5790-4d90-b351-8b87ebfbbe3a) |
<a name="f370a751-0004-4c86-8d5d-26597800b5c4"></a>
### 8419-e15c658/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31200
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fd43f2045d02:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ajpmoawkmjfvsd",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365832,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15CEF315DB5800AB, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="ddd6ad3c-e133-4435-8206-e64f6ee41bb7"></a>
### 8419-e15c658/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31164
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 00c7c60945b2:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 55 seconds
(4/14) Running aws-sdk-ruby tests ... done in 18 seconds
(5/14) Running awscli tests ... done in 1 minutes and 21 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 11 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 35 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 56 seconds
{
  "args": {},
  "duration": 93601,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="6d0f7539-5790-4d90-b351-8b87ebfbbe3a"></a>
### 8419-e15c658/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31182
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 26a1bb648737:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 6 seconds
(5/14) Running awscli tests ... done in 2 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 51 seconds
(8/14) Running minio-dotnet tests ... done in 37 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 14 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 45 seconds
(11/14) Running minio-js tests ... done in 54 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 25 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1396, in test_presigned_get_object_response_headers\n    returned_content_language = response.headers['Content-Language']\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/_collections.py\", line 156, in __getitem__\n    val = self._container[key.lower()]\nKeyError: 'content-language'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2039, in main\n    test_presigned_get_object_response_headers(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1414, in test_presigned_get_object_response_headers\n    raise Exception(err)\nException: 'content-language'\n",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "message": "'content-language'",
  "name": "minio-py:test_presigned_get_object_response_headers",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-b7bf009a-853c-4e77-ab97-b214348f908b",
    "object_name": "6fb64766-58a9-41dc-873f-f0bbff443b2c",
    "presigned_get_object_url": "http://72.28.97.56:31182/minio-py-test-b7bf009a-853c-4e77-ab97-b214348f908b/6fb64766-58a9-41dc-873f-f0bbff443b2c?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191019%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191019T045531Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=f2a064288c54946ffb6f59b2463cbb25d9100e0035336780f60ab9f974eef79d"
  },
  "duration": 98
}

Executed 11 out of 14 tests successfully.
```
<a name="fbe03e75-78c0-4df0-a342-f046141b1089"></a>
### 8419-e15c658/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.58:32723
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 00b2ca5f1278:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-jdp9nxyu0mo1vn",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365699,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15CEF315DDB9EAA1, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
Error: No such image: minio/minio:8419-e15c658

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#3158bc6d-631d-4676-95a6-92433b93a60d) |
<a name="3158bc6d-631d-4676-95a6-92433b93a60d"></a>
### 8419-c2431a9/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31290
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5609a0f52f59:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 59 seconds
(4/14) Running aws-sdk-ruby tests ... done in 12 seconds
(5/14) Running awscli tests ... done in 2 minutes and 25 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 14 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 34 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 0 seconds
{
  "args": {},
  "duration": 85549,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8419-c2431a9

**Review by vadmeste — APPROVED** [review]
LGTM & tested

File: cmd/xl-sets.go [hunk 1]
```diff
@@ -1711,23 +1638,37 @@ func (s *xlSets) ListBucketsHeal(ctx context.Context) ([]BucketInfo, error) {
...
 
-	endWalkCh := make(chan struct{})
-	listDir := listDirSetsFactory(ctx, s.sets...)
-	walkResultCh := startTreeWalk(ctx, bucket, prefix, "", recursive, listDir, endWalkCh)
+	marker := ""
 	for {
-		walkResult, ok := <-walkResultCh
-		if !ok {
-			break
+		if globalHTTPServer != nil {
```
Comment (vadmeste) [inline cmd/xl-sets.go:1645]: @harshavardhana, we do not need to pause healing here, this is already done in background-heal-ops.go

**Review by vadmeste — COMMENTED** [review]
One comment

**Review by kannappanr — APPROVED** [review]
LGTM

File: cmd/xl-sets.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/xl-sets.go:1645]: Listing is slowing down itself further if needed @vadmeste - we need to comprehensively fix this by implementing suspend, resume style for these type of calls.

**Pull request merged**